server/lib/chat/chat-models.ts

```diff
@@ -578,3 +580,116 @@ export async function findChannelByName(
     done()
   }
 }
+
+/**
+ * Returns a list of non-private chat channels, optionally filtered by a `searchStr`. The list
+ * doesn't include the user's joined channel by default, but can be configured to include them as
+ * well.
+ */
+export async function listChannels(
+  {
+    userId,
+    limit,
+    pageNumber,
+    searchStr,
+    joined,
+  }: {
+    /** User requesting the list of channels. */
+    userId: SbUserId
+    /** Amount of channels to return. */
+    limit: number
+    /** Offset from which to start listing the channels. */
+    pageNumber: number
+    /** String to filter the channels by. */
+    searchStr?: string
+    /** Flag which includes the user's joined channels if true. */
+    joined?: boolean
+  },
+  withClient?: DbClient,
+): Promise<ChannelInfo[]> {
+  const { client, done } = await db(withClient)
+  try {
+    const query = joined
+      ? sql`
+        SELECT c.*
+        FROM channels c
+      `
+      : sql`
+        WITH unjoined_channels AS (
+          SELECT DISTINCT(channel_id)
+          FROM channel_users
+          WHERE channel_id NOT IN (SELECT channel_id FROM channel_users WHERE user_id = ${userId})
+        )
+        SELECT c.*
+        FROM channels c
+        INNER JOIN unjoined_channels ON c.id = unjoined_channels.channel_id
+      `
+
+    const whereCondition = sql`WHERE private = false`
```

Comment:
I thought about it, and I'm not sure if we ever want to display private channels in the search results for non-admins, even the ones you've joined in. So there are two use cases I was thinking of where we might use this API:
- join-channel dialog: here we'll probably use it without the `joined` flag, so it doesn't matter
- in the autocomplete box when mentioning the channel in chat: here we'll probably use the `joined` flag (and order them on top of results), but I don't think we want to show the private channels you're in when trying to mention a channel

Reply:
When you're writing external APIs, it's generally more okay to have them targeted to particular usecases, mostly because REST is not particularly expressive as a type system and the effort required to serve a general-purpose API that we will never use is not worth it.

When you're writing APIs for internal usage (which is what all of our model APIs are), they should either be properly limited by the types+parameters they allow, or as general as the types+parameters imply. The parameters you've created here, to me, imply that I will get all the channels that I've joined, regardless of whether or not they are private. This system does not imply that we are limited to 2 usecases only, and does not imply the assumptions you've stated above. If you want a system that does that, you probably want to create 2 model methods instead, one for each usecase, that are named specifically for those usecases. Otherwise, someone is going to come along 2 months from now and use this method thinking it does what they want, when it very much does not, we're not going to notice, it'll get shipped into production, and 6 months down the line we'll realize it's been broken the whole time.

As far as the usecases you stated above, I think your second one is also just mistaken, I don't see why we would ever want to avoid having a joined channel in the mention autocomplete. The channels are just not able to be joined without an invite, they're not hidden, and that has never been the meaning of that flag.